### PR TITLE
refactor: migrate Encryptor to TypeScript and increase PBKDF2 iterations number

### PR DESCRIPTION
**Pull request by gantunesr**
## **Description**

📣 **The code in this PR has already been reviewed and QA'd in https://github.com/MetaMask/metamask-mobile/pull/8828**

This PR introduces the following modifications

- `@metamask/keyring-controller` bump from `v8.1.0` to `v9.0.0`
- Refactor the Encryptor class from JS to TS
- Adds unit tests to the Encryptor class
- Increases the number of iterations for PBKDF2 from 5.000 to 900.000 

### CHANGELOG from `@metamask/keyring-controller`

> Added,
> - Add `KeyringController:persistAllKeyrings` messenger action ([#1965](https://github.com/MetaMask/core/pull/1965))
> 
> Changed,
> - **BREAKING** Change `encryptor` constructor option property type to `GenericEncryptor | ExportableKeyEncryptor | > undefined` ([#2041](https://github.com/MetaMask/core/pull/2041))
>   - When the controller is instantiated with `cacheEncryptionKey: true`, `encryptor` may no longer be of type `GenericEncryptor`.
> - Bump dependency on `@metamask/scure-bip39` 2.1.1 ([#1868](https://github.com/MetaMask/core/pull/1868))
> - Bump dependency on `@metamask/utils` to 8.2.0 ([#1957](https://github.com/MetaMask/core/pull/1957))
> - Bump @metamask/eth-keyring-controller to 14.0.0 ([#1771](https://github.com/MetaMask/core/pull/1771))

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/100

## **Manual testing steps**

<details>
  <summary>Test case 1: <b>Upgrade the client from the current version to feature branch</b></summary>
  <br>
    <ol>
      <li>Install the current version of MM mobile or the last commit in the main branch</li>
      <li>Create or import a wallet and verify it is correctly created</li>
      <li>Lock the wallet</li>
      <li>Upgrade the app to this feature branch</li>
      <li>Unlock the wallet, it should unlock the wallet</li>
    </ol>
</details>

<details>
  <summary>Test case 2: <b>Lock and unlock the wallet</b></summary>
  <br>
    <ol>
      <li>Create or import a wallet and verify it is correctly created</li>
      <li>Lock the wallet</li>
      <li>Unlock the wallet, it should unlock the wallet</li>
    </ol>
</details>

<details>
  <summary>Test case 3: <b>Reveal SRP and private key</b></summary>
  <br>
    <ol>
      <li>Create or import a wallet and verify it is correctly created</li>
      <li>Go to Settings > Security & Privacy</li>
      <li>Reveal the SRP, it should show the correct SRP</li>
      <li>Go back to Security & Privacy</li>
      <li>Reveal the private key of the current account</li>
    </ol>
</details>

<details>
  <summary>Test case 4: <b>Connect QR wallet</b></summary>
  <br>
    <ol>
      <li>Create or import a wallet and verify it is correctly created</li>
      <li>Connect QR wallet and import 2 or more accounts</li>
    </ol>
</details>

<details>
  <summary>Test case 5: <b>Connect Ledger wallet</b></summary>
  <br>
    <ol>
      <li>Create or import a wallet and verify it is correctly created</li>
      <li>Connect Ledger wallet and import the account</li>
    </ol>
</details>

<details>
  <summary>Test case 6: <b>Sign a transaction with each account type</b></summary>
  <br>
    <ol>
      <li>Create or import a wallet and verify it is correctly created</li>
      <li>Connect Ledger wallet and import the account</li>
      <li>Connect QR wallet and import 2 or more accounts</li>
      <li>Send a transaction in a test network using each account type (HD, QR, and Ledger)</li>
    </ol>
</details>

<details>
  <summary>Test case 7: <b>Sign a different messages with each account type</b></summary>
  <br>
    <ol>
      <li>Create or import a wallet and verify it is correctly created</li>
      <li>Connect Ledger wallet and import the account</li>
      <li>Connect QR wallet and import 2 or more accounts</li>
      <li>Go to <a href="[url](https://metamask.github.io/test-dapp/)">MM E2E Test Dapp</a> and sign different messages (Eth Sign, Personal Sign, Sign Typed Data, Sign Typed Data V3, Sign Typed Data V4, Sign Permit, Sign In With Ethereum) with each account (HD, QR, and Ledger). Note: Some methods are not supported in Ledger</li>
    </ol>
</details>

<details>
  <summary>Test case 8: <b>Change password</b></summary>
  <br>
    <ul>
      <li><a href="https://www.notion.so/metamask-consensys/Change-Password-7a8a5cbfa60b409ea3527214baa14e00">Document with test case</></li>
    </ul>
</details>

<details>
  <summary>Test case 9: <b>Unlock with biometrics</b></summary>
  <br>
    <ul>
      <li><a href="https://www.notion.so/metamask-consensys/Unlock-with-Biometrics-283588545d0e419ca3d5e0ac98f2e35d">Document with test case</></li>
    </ul>
</details>

<details>
  <summary>Test case 10: <b>Remember me</b></summary>
  <br>
    <ul>
      <li><a href="https://www.notion.so/metamask-consensys/Remember-Me-RM-6783c5a6dd3f498e8a75886a53269659">Document with test case</></li>
    </ul>
</details>

<details>
  <summary>Test case 11: <b>Onboarding with biometrics</b></summary>
  <br>
    <ul>
      <li><a href="https://www.notion.so/metamask-consensys/Biometrics-for-Onboarding-Face-ID-or-Fingerprint-ad1162c9111c443a8f38027039fd3e1a">Document with test case</></li>
    </ul>
</details>

## **Unit Tests and Coverage**

```
yarn jest -u app/core/Encryptor/Encryptor.test.ts --collectCoverage --collectCoverageFrom=app/core/Encryptor/Encryptor.ts
```

```
=============================== Coverage summary ===============================
Statements   : 100% ( 46/46 )
Branches     : 94.73% ( 18/19 )
Functions    : 100% ( 14/14 )
Lines        : 100% ( 45/45 )
================================================================================
```

## **Screenshots/Recordings**

Not applicable. No changes to the UI/UX.

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - In case it's not yet "ready for review", I've set it to "draft".
  - In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by gantunesr** [comment]
@SocketSecurity ignore npm/@metamask/eth-keyring-controller@15.1.0

**Review by tommasini — DISMISSED** [review]
LGTM! Awesome refactor on the Encryptor class

File: app/core/Encryptor/Encryptor.ts [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    btoa(String.fromCharCode.apply(null, Array.from(byteArray)));
+
+  /**
+   * Wrapper method for key generation from a password.
+   * @param params.password - The password used for key derivation.
+   * @param params.salt - The salt used for key derivation.
+   * @param params.lib - The library to use ('original' or forked version).
+   * @returns A promise that resolves to the derived encryption key.
+   */
+  private generateKeyFromPassword = ({
```
Comment (FrederikBolding) [inline app/core/Encryptor/Encryptor.ts:110]: Could this match the `browser-passworder` API closer? Specifically exposing `keyFromPassword`, `importKey`, `exportKey`, `encryptWithKey`, `decryptWithKey`? Ideally with as close as identical API as possible.

That would make it easier to support something like: https://github.com/MetaMask/snaps/blob/694c3677bfb33fdbc616af55feab7d4955604c08/packages/snaps-controllers/src/snaps/SnapController.ts#L1561-L1568

Might let us simplify the KeyringController too

Comment (gantunesr) [reply app/core/Encryptor/Encryptor.ts:110]: Good point, let me get that done today

Comment (gantunesr) [reply app/core/Encryptor/Encryptor.ts:110]: This request is being addressed in this PR

https://github.com/MetaMask/metamask-mobile/pull/9203

File: app/core/Encryptor/Encryptor.test.ts [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+        description:
+          'with original library and old iterations number for key generation',
+      },
+      {
+        lib: 'random-lib', // Assuming not using "original" should lead to AesForked
+        expectedKey: 'mockedAesForkedKey',
+        expectedPBKDF2Args: ['testPassword', 'mockedSalt'],
+        description:
+          'with library different to "original" and default iterations number for key generation',
+        keyMetadata: DEFAULT_DERIVATION_PARAMS,
```
Comment (ccharly) [inline app/core/Encryptor/Encryptor.test.ts:109]: After doing some refactor to align with `browser-passworder` I found out that this test is "incorrect".

The `AesForked` library actually uses a fixed number of round:
- https://github.com/MetaMask/react-native-aes/blob/master/android/src/main/java/com/tectiv3/aes/RCTAes.java#L34
- https://github.com/MetaMask/react-native-aes/blob/master/ios/RCTAes/lib/AesCrypt.m#L62

Ultimately, we could raise an error if the `keyMetadata` parameters does not match what is being used "underneath".

It has no real impact right now, but just wanted to point that out!

File: app/core/Encryptor/Encryptor.ts [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+  private encryptWithKey = async ({
+    text,
+    keyBase64,
+  }: {
+    text: string;
+    keyBase64: string;
+  }): Promise<EncryptionResult> => {
+    const iv = await Aes.randomKey(16);
+    return Aes.encrypt(text, keyBase64, iv).then((cipher: string) => ({
+      cipher,
```
Comment (ccharly) [inline app/core/Encryptor/Encryptor.ts:140]: Small note, `EncryptionResult` has no `cipher` field (its named `data` instead. It looks like ts is ok with that (which is a bit surprising...).

I did fix that in the next PR: https://github.com/MetaMask/metamask-mobile/pull/9203

Up to you if you wanna fix the type! :)

File: app/core/Encryptor/types.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+ * The result of an encryption operation.
+ * @interface EncryptionResult
+ * @property data - The encrypted data.
+ * @property iv - The initialization vector used in the encryption process.
+ * @property [salt] - The salt used in the encryption process, if applicable.
+ * @property [lib] - The library or algorithm used for encryption, if applicable.
+ * @property [keyMetadata] - Metadata about the key derivation, if key derivation was used.
+ */
+export interface EncryptionResult {
+  data: string;
```
Comment (ccharly) [inline app/core/Encryptor/types.ts:33]: (Related to my previous comment) This should be `cipher` instead.

**Review by ccharly — DISMISSED** [review]

**Review by tommasini — DISMISSED** [review]
LGTM! Thanks for decreasing the number of iterations! Curious question, was it because of performance?

**Comment by gantunesr** [comment]
> _Curious question, was it because of performance?_

@tommasini Yes, it is for performance reasons

**Review by owencraston — DISMISSED** [review]

**Review by tommasini — APPROVED** [review]

**Review by ccharly — APPROVED** [review]
LGTM. Just left minor remarks that we can handle later on! :)

File: app/core/Encryptor/Encryptor.test.ts [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+      const vault = JSON.parse(updatedVault);
+
+      expect(encryptSpy).toBeCalledTimes(1);
+      expect(decryptSpy).toBeCalledTimes(1);
+      expect(vault).toHaveProperty('keyMetadata');
+      expect(vault.keyMetadata).toStrictEqual(expectedKeyMetadata);
+    });
+
+    it('does not update a vault if algorithm is PBKDF2 and the number of iterations is 900000', async () => {
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/core/Encryptor/Encryptor.test.ts:221]: I'll address this comment in a follow up PR

File: app/core/Encryptor/Encryptor.test.ts [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+  describe('constructor', () => {
+    it('throws an error if the provided iterations do not meet the minimum required', () => {
+      expect(
+        () =>
+          new Encryptor({
+            derivationParams: {
+              algorithm: 'PBKDF2',
+              params: {
+                iterations: 100,
```
[the remark this answers is not in the document]
Comment (gantunesr) [reply app/core/Encryptor/Encryptor.test.ts:27]: I'll address this comment in a follow up PR

**Pull request merged**